crates/interpreter/src/valid.rs

```diff
@@ -26,25 +27,33 @@ use crate::util::*;
 use crate::*;
 
 /// Checks whether a WASM module in binary format is valid.
-pub fn validate(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
-    Context::default().check_module(&mut Parser::new(binary))
+pub fn validate<M: ValidMode>(binary: &[u8]) -> Result<Vec<MetadataEntry>, Error> {
+    Context::<Prepare>::default().check_module(&mut Parser::new(binary))
```

Comment:
I guess it's better to do the design first, like on #726. For this PR, since there's only `Prepare`, I would suggest not making `validate` generic. Only `Context` and its methods need to be. We would then have 2 public functions (and `ValidMode` can stay private):

```rust
pub fn prepare(binary: &[u8]) -> Result<Vec<u8>, Error>; // returns the sidetable section
pub fn validate(binary: &[u8]) -> Result<(), Error>; // the binary has a sidetable section in this case
```

We could even have:

```rust
pub fn prepare(binary: &mut Vec<u8>) -> Result<(), Error>; // appends the sidetable section
```